torchtext/csrc/vocab.cpp

```diff
@@ -0,0 +1,142 @@
+
+public:
+  // stoi_, and unordered_map holds the serialized params passed in
+  // during initialization. We need this because we need to be able to serialize
+  // the model so that we can save the scripted object. Pickle will get the
+  // serialized model from these members, thus they needs to be public.
+  std::vector<std::string> itos_;
+  std::string unk_token_;
+
+  explicit Vocab(const std::vector<std::string> &tokens,
+                 const std::string &unk_token)
+      : itos_(std::move(tokens)), unk_token_(std::move(unk_token)) {
+    stoi_.reserve(tokens.size());
+    for (std::size_t i = 0; i < tokens.size(); i++) {
+      // tokens should not have any duplicates
+      if (stoi_.find(tokens[i]) != stoi_.end()) {
+        throw std::runtime_error("Duplicate token found in tokens list: " +
+                                 tokens[i]);
+      }
+      stoi_.insert(std::move(tokens[i]), i);
+    }
+    unk_index_ = stoi_.find(unk_token)->value();
+  }
+
+  int64_t __len__() const { return stoi_.size(); }
+
+  int64_t __getitem__(const std::string &token) const {
+    const auto &item = stoi_.find(token);
+    if (item != stoi_.end()) {
+      return item->value();
+    }
+    return unk_index_;
+  }
+
+  void __setitem__(const std::string &token, const int64_t &index) {
+    if (index < 0 || index > static_cast<int64_t>(stoi_.size())) {
+      throw std::runtime_error(
+          "Specified index " + std::to_string(index) +
+          " is out of bounds of the size of stoi dictionary: " +
+          std::to_string(stoi_.size()) + ".");
+    }
+
+    const auto &item = stoi_.find(token);
+    // if item already in stoi we throw an error
+    if (item != stoi_.end()) {
+      throw std::runtime_error("Token " + token +
+                               " already exists in the Vocab with index: " +
+                               std::to_string(item->value()) + ".");
+    }
+
+    // need to offset all tokens greater than or equal index by 1
+    for (auto &entry : stoi_) {
+      if (entry.value() >= index) {
+        stoi_.insert_or_assign(entry.key(), std::move(entry.value() + 1));
+      }
+    }
+    stoi_.insert(std::move(token), std::move(index));
+
+    // need to update unk_index in case token equals unk_token or token inserted
+    // before unk_token
+    unk_index_ = stoi_.find(unk_token_)->value();
+  }
+
+  void addToken(const std::string &token) {
+    if (stoi_.find(token) == stoi_.end()) {
+      stoi_.insert(std::move(token), stoi_.size());
+    }
+  }
+
+  std::string lookupToken(const int64_t &index) {
+    if (index < 0 || index > static_cast<int64_t>(itos_.size())) {
+      throw std::runtime_error(
+          "Specified index " + std::to_string(index) +
+          " is out of bounds of the size of itos dictionary: " +
+          std::to_string(itos_.size()) + ".");
+    }
+
+    return itos_[index];
+  }
+
+  std::vector<std::string> lookupTokens(const std::vector<int64_t> &indices) {
```

Comment:
Not sure if `itos` is really necessary, especially a list of index to a list of tokens.

Reply:
So the `itos` is necessary because when we are serializing a Vocab we need to somehow maintain the ordering for all the tokens (this is currently done by the `itos` vector). Without the `itos` we wouldn't be able to deserialize the saved model in the correct order.